README.rst
Health check for PostgreSQL + repmgr
====================================

This package is intended to be used as an health check backend for HAProxy. It
provides a small script that will run an HTTP server querying each PostgreSQL
servers of the replication cluster to know the role (primary, standby, witness
or fenced) of the local instance according to the majority of them (kind of a
vote).

Usage and configuration example to come…

TODO: make use of Tornado for HTTP server
TODO: make use of asyncio for SQL queries
TODO: split HTTP server and SQL queries in two modules to allow running SQL part through a CLI program
TODO: replace json by toml (with backward compatibility)
TODO: use poetry for dependencies management and packaging
TODO: handle environment variables for config (with os.environ.get(x))



CHANGELOG.md
# CHANGELOG

## v0.4.2

* remove logging and let journalctl do its job

## v0.4.1

* fix psql connect by merging dicts

## v0.4.0

* make psycopg2 timeout configurable
* add basic logging

## v0.3.1

* set MIME type to text/plain

## v0.3.0

* add ipv6 compatibility

## v0.2.0

* allow to configure listening address
* allow to configure postgresql target port

## v0.1.0

* first release


hapcheck.py
#!/usr/bin/env python

from __future__ import (absolute_import, division,
                        print_function, unicode_literals)
from builtins import *

import http.server
import psycopg2

PORT = 8000
NODES = {
    'pg1': {
        'addr': '10.64.249.79',
        'node_id': 1,
    },
    'pg2': {
        'addr': '10.64.249.188',
        'node_id': 2,
    },
    'pg3': {
        'addr': '10.64.249.155',
        'node_id': 3,
    },
}
CONN = {
    'dbname': 'repmgr',
    'user': 'repmgr',
    'password': 'changeme',
}


class Server(http.server.HTTPServer):
    pass


class Handler(http.server.SimpleHTTPRequestHandler):
    def __init__(self, *args, **kwargs):
        self.s = get_quorum_state(**NODES['pg1'])
        http.server.SimpleHTTPRequestHandler.__init__(self, *args, **kwargs)

    def _set_headers(self):
        self.send_response(200)
        self.send_header('Content-Type', 'application/json')
        self.end_headers()

    def do_GET(self):
        self._set_headers()
        self.wfile.write(self.s)
        self.wfile.write('\n')

    def do_HEAD(self):
        self._set_headers()


def get_state(addr, node_id):
    try:
        # postgresql query
        con = psycopg2.connect(host=addr, **CONN)
        cur = con.cursor()
        query = 'SELECT active, type FROM repmgr.nodes WHERE node_id = {};'
        cur.execute(query.format(node_id))
        data = cur.fetchone()
        cur.close()
        # determine if is an active primary
        is_primary = True if data == (True, 'primary') else False
        # return result
        return is_primary
    except Exception:
        # an error occured, so return false by default
        return False


def get_quorum_state(addr, node_id):
    # init vars
    vote = 0
    state = 'standby'
    # ask each node for its state
    for node in NODES.values():
        vote += 1 if get_state(node['addr'], node_id) else 0
    # determines voting result
    if vote > len(NODES) / 2:
        state = 'primary'
    # return result
    return state


def run():
    # run http server
    httpd = Server(('localhost', PORT), Handler)
    print('serving at port', PORT)
    httpd.serve_forever()


if __name__ == '__main__':
    run()


src/rephacheck/server.py
#!/usr/bin/env python
# -*- encoding: utf-8 -*-

"""
Determine by voting which is the state of each node.
For this to work properly, you need to have an odd number of nodes.
"""

from __future__ import (absolute_import, division,
                        print_function, unicode_literals)
from builtins import *

from collections import Counter
import http.server
from io import open
import json
import psycopg2
from socket import AF_INET6
import sys

with open('/etc/rephacheck.json') as rephaconf:
    conf = json.load(rephaconf)

LISTEN = conf.get('listen', 'localhost')
PORT = conf.get('port', 8000)
NODES = conf.get('nodes')
CONNINFO = conf.get('conninfo')
TIMEOUT = conf.get('timeout', 5)
CURRENT = conf.get('local_node_id')


class Server(http.server.HTTPServer):
    address_family = AF_INET6


class Handler(http.server.SimpleHTTPRequestHandler):
    def __init__(self, *args, **kwargs):
        self.s = get_quorum_state(CURRENT)
        http.server.SimpleHTTPRequestHandler.__init__(self, *args, **kwargs)

    def _set_headers(self):
        self.send_response(200)
        self.send_header('Content-Type', 'text/plain')
        self.end_headers()

    def do_GET(self):
        self._set_headers()
        self.wfile.write(self.s.encode('utf-8'))
        self.wfile.write('\n'.encode('utf-8'))

    def do_HEAD(self):
        self._set_headers()

    def log_message(self, format, *args):
        sys.stderr.write("%s - %s\n" % (self.client_address[0], format % args))

    def log_request(self, code='-', size='-'):
        self.log_message('"%s" %s %s %s',
                         self.requestline, str(code), str(size), self.s)


def get_state(addr, port, node_id):
    try:
        # postgresql query
        con_args = {'host': addr, 'port': port, 'connect_timeout': TIMEOUT}
        con_args.update(CONNINFO)
        con = psycopg2.connect(**con_args)
        cur = con.cursor()
        query = 'SELECT active, type FROM repmgr.nodes WHERE node_id = {};'
        cur.execute(query.format(node_id))
        data = cur.fetchone()
        cur.close()
        # return result
        return data
    except Exception:
        # an error occured, so return false by default
        return (False, 'unknown')


def get_quorum_state(node_id):
    # init vars
    votes = []
    # ask each node for the state of `node_id`
    for node in NODES.values():
        active, role = get_state(node['addr'], node['port'], node_id)
        # if node considered active take vote, otherwise fence it
        if active:
            votes.append(role)
        else:
            votes.append('fenced')
    # determines voting result
    results = Counter(votes)
    state = results.most_common(1)[0][0]
    # return result
    return state


def run():
    # run http server
    httpd = Server((LISTEN, PORT), Handler)
    print('serving at port', PORT)
    httpd.serve_forever()


if __name__ == '__main__':
    run()
